### PR TITLE
Fix race on sending stdin close event

### LINKED ISSUE #28658: stdin not closed on 1.13.0-rc1
Create a largeish file (60M is good) and feed it into a container:
```
dd if=/dev/zero of=file bs=1M count=60
tar cf - file | docker run -i --log-driver=none alpine tar xfv -
```
Run the `docker run` line a few times and it will freeze, signals won't work and the process will just be stuck. Container is still there. Have to restart daemon to fix.

You don't need `--log-driver=none` to repro, or the `tar` eg
```
dd if=/dev/zero bs=1M count=60 | docker run -i alpine sh -c 'cat > /dev/null'
```

will repro just fine too.

**Additional information you deem important (e.g. issue happens only occasionally):**

Can repro after a very short number of tries (1-3)

**Output of `docker version`:**

```
Client:
 Version:      1.13.0-rc1
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   75fd88b
 Built:        Fri Nov 11 19:47:07 2016
 OS/Arch:      linux/amd64

Server:
 Version:             1.13.0-rc1
 API version:         1.25
 Minimum API version: 1.12
 Go version:          go1.7.3
 Git commit:          75fd88b
 Built:               Fri Nov 11 19:47:07 2016
 OS/Arch:             linux/amd64
 Experimental:        false
```

Also happens on current `1.13.x` branch, and `master`

**Output of `docker info`:**

```
Containers: 3
 Running: 1
 Paused: 0
 Stopped: 2
Images: 1
Server Version: 1.13.0-rc1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 7
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 8517738ba4b82aff5662c97ca4627e7e4d03b531
runc version: ac031b5bf1cc92239461125f4c1ffb760522bbf2
init version: N/A (expected: v0.13.0)
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-47-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.953 GiB
Name: hang-rc1
ID: AJEZ:COTX:SQHN:V5NJ:BBZT:TNGS:54EF:U3ES:JQ6R:H2N7:QJYR:H5D3
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Digital Ocean clean Ubuntu VM, also repro on Docker for Mac master branch.

### PR DESCRIPTION
**Pull request by tonistiigi**
fixes #28658

Close event could have been ignored when it happens late and both `ready` and `ctx.Done` channels are closed.

@cpuguy83 @justincormack 

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

### CODE REVIEW

**Comment by justincormack** [comment]
Yes, fixes it for me.

Might be nice to add a regression test, eg just `dd if=/dev/zero bs=1M count=60 | docker run -i busybox sh -c 'cat > /dev/null'`

**Comment by tonistiigi** [comment]
@justincormack Test added

**Review by justincormack — APPROVED** [review]

**Review by vdemeester — APPROVED** [review]
LGTM :frog: 
/cc @LK4D4

File: integration-cli/docker_cli_run_test.go [hunk 1]
```diff
@@ -4849,3 +4850,34 @@ func (s *DockerSuite) TestRunEmptyEnv(c *check.C) {
 	c.Assert(err, checker.NotNil)
 	c.Assert(out, checker.Contains, expectedOutput)
 }
+
+// #28658
+func (s *DockerSuite) TestSlowStdinClosing(c *check.C) {
+	testRequires(c, DaemonIsLinux)
```
Comment (AkihiroSuda) [inline integration-cli/docker_cli_run_test.go]: Can we enable this test for Windows as well, although it is not related to the fix for container_unix.go?

Comment (tonistiigi) [inline integration-cli/docker_cli_run_test.go]: What binary can I use to block on stdin in windows?

Comment (AkihiroSuda) [inline integration-cli/docker_cli_run_test.go]: IIUC, you can just use the busybox cat here in Windows
https://github.com/docker/docker/blob/420e22fad4dcdff3b8025666216cc85fc9358f69/integration-cli/docker_cli_inspect_test.go#L145

**Comment by AkihiroSuda** [comment]
CI failure seems unrelated, restarting
```
01:45:14.904 PASS: docker_cli_pull_test.go:194: DockerHubPullSuite.TestPullAllTagsFromCentralRegistry	Agent went offline during the build
01:45:14.914 Build step 'Execute shell' marked build as failure
01:45:14.942 34.439sSet build name.
01:45:14.942 Workspace is not accessible
01:45:14.963 
01:45:14.965 PASS: docker_cli_pull_test.go:237: DockerHubPullSuite.TestPullClientDisconnect	3.287s
01:45:14.968 [PostBuildScript] - Execution post build scripts.
01:45:14.969 ERROR: Build step failed with exception
01:45:14.969 java.lang.NullPointerException: no workspace from node hudson.plugins.ec2.EC2OndemandSlave[ubuntu-1404 (aufs) (i-8e32b01b)] which is computer hudson.plugins.ec2.EC2Computer@2698235b and has channel null
```

File: integration-cli/docker_cli_run_test.go [hunk 2]
```diff
@@ -4849,3 +4850,34 @@ func (s *DockerSuite) TestRunEmptyEnv(c *check.C) {
...
+
+	name := "testslowstdinclosing"
+	repeat := 3 // regression happened 50% of the time
+	for i := 0; i < repeat; i++ {
+		cmd := exec.Command(dockerBinary, "run", "--rm", "--name", name, "-i", "busybox", "cat")
```
Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: minor nit, but can use `exec.CommandContext`

Comment (tonistiigi) [inline integration-cli/docker_cli_run_test.go]: are we `go1.7` only now?

Comment (tonistiigi) [inline integration-cli/docker_cli_run_test.go]: also, doesn't work with `runCommand` implementation
